hax/treemakers/posrec.py

```diff
@@ -0,0 +1,200 @@
+            "x_observed_nn_tf": None,
+            "y_observed_nn_tf": None,
+            "s1_area_upper_injection_fraction": None,
+            "s1_area_lower_injection_fraction": None,
+        }
+
+        # We first need the positions. This minitree is only valid when loading
+        # Corrections since you need that to get the corrected positions
+        if not len(event.interactions):
+            return event_data
+        e = event.event_number
+        try:
+            i = self.indices.index(e)
+        except Exception:
+            return event_data
+        interaction = event.interactions[0]
+        s1 = event.peaks[interaction.s1]
+        s2 = event.peaks[interaction.s2]
+
+        # Position reconstruction based on NN from TensorFlow
+        s2apc = np.array(list(s2.area_per_channel))
+        s2apc_clean = []
+        for i, s2_t in enumerate(s2apc):
+            if i not in self.list_bad_pmts and i < self.ntop_pmts:
+                s2apc_clean.append(s2_t)
+        s2apc_clean = np.asarray(s2apc_clean)
+        s2apc_clean_norm = s2apc_clean / s2apc_clean.sum()
+        s2apc_clean_norm = s2apc_clean_norm.reshape(1, len(s2apc_clean_norm))
+        predicted_xy_tensorflow = self.nn_tensorflow.predict(s2apc_clean_norm)
+        event_data['x_observed_nn_tf'] = predicted_xy_tensorflow[0, 0] / 10.
+        event_data['y_observed_nn_tf'] = predicted_xy_tensorflow[0, 1] / 10.
+        
+        # s1 area fraction near injection points for Rn220 source
+        area_upper_injection = s1.area_per_channel[131] + s1.area_per_channel[138] + s1.area_per_channel[146] + s1.area_per_channel[147]
+        area_lower_injection = s1.area_per_channel[236] + s1.area_per_channel[237] + s1.area_per_channel[243] 
+        
+        event_data['s1_area_upper_injection_fraction'] = area_upper_injection / s1.area
+        event_data['s1_area_lower_injection_fraction'] = area_lower_injection / s1.area
+        
+        # Want S1 AreaFractionTop Probability
+        if s1.area < self.low_pe_threshold:
+            s1_frac = s1.area / self.low_pe_threshold
+            hits_top = s1.n_hits * s1.hits_fraction_top
+            s1_top = s1.area * s1.area_fraction_top
+            size_top = hits_top * (1. - s1_frac) + s1_top * s1_frac
+            size_tot = s1.n_hits * (1. - s1_frac) + s1.area * s1_frac
+        else:
+            size_top = s1.area * s1.area_fraction_top
+            size_tot = s1.area
+
+        aft = self.aft_map.get_value(self.x[i], self.y[i], self.z[i])
+        event_data['s1_area_fraction_top_probability_hax'] = binom_test_pax(
+            size_top, size_tot, aft)
+
+        # Now do s1_pattern_fit
+        apc = np.array(list(s1.area_per_channel))
+        hpc = np.array(list(s1.hits_per_channel))
+
+        # Get saturated channels
+        confused_s1_channels = []
+        for a, c in enumerate(s1.n_saturated_per_channel):
+            if c > 0:
+                confused_s1_channels.append(a)
```

Comment:
Can be one line https://github.com/XENON1T/pax/blob/master/pax/datastructure.py#L249